[Concurrency] Eliminate StatusRecordLockRecord.

mikeash opened the pull request:
Draft PR, this still needs a Win32 implementation of `RecursiveMutex`.

Move to a recursive lock inline in the Task. This avoids the need to allocate a lock record and simplifies the code somewhat.

Change Task's OpaquePrivateStorage to compute its size at build time based on the sizes of its components, rather than having it be a fixed size. It appears that the fixed size was intended to be part of the ABI, but that didn't happen and we're free to change this size. We need to expand it slightly when using pthread_mutex as the recursive lock, as pthread_mutex is pretty big. Other recursive locks allow it to shrink slightly.

We don't have a recursive mutex in our Threading support code, so add a RecursiveMutex type.

rdar://113898653

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

mikeash committed:
[Concurrency] Eliminate StatusRecordLockRecord.

Move to a recursive lock inline in the Task. This avoids the need to allocate a lock record and simplifies the code somewhat.

Change Task's OpaquePrivateStorage to compute its size at build time based on the sizes of its components, rather than having it be a fixed size. It appears that the fixed size was intended to be part of the ABI, but that didn't happen and we're free to change this size. We need to expand it slightly when using pthread_mutex as the recursive lock, as pthread_mutex is pretty big. Other recursive locks allow it to shrink slightly.

We don't have a recursive mutex in our Threading support code, so add a RecursiveMutex type.

rdar://113898653

ktoso reviewed (approved):
Looks great! Great to use a plain old lock rather than our own record lock dance there, much nicer.

mikeash commented:
@swift-ci Please Build Toolchain Linux Platform

mikeash commented:
@swift-ci Please Build Toolchain macOS Platform

al45tair committed:
[Concurrency] Add Win32 implementation for recursive mutex.

We're using `CRITICAL_SECTION` here, as that supports recursion.

rdar://113898653

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented on include/swift/Threading/Mutex.h:
```diff
@@ -241,6 +241,48 @@ class IndirectMutex {
+/// A recursive variant of Mutex.
+class RecursiveMutex {
+
+  RecursiveMutex(const Mutex &) = delete;
+  RecursiveMutex &operator=(const Mutex &) = delete;
+  RecursiveMutex(Mutex &&) = delete;
+  RecursiveMutex &operator=(Mutex &&) = delete;
```
I think these want to take `RecursiveMutex` rather than `Mutex` arguments.

mikeash replied:
Copy/pasta trouble, fixed all of these now.

al45tair commented on include/swift/Threading/Impl/Linux.h:
```diff
@@ -134,6 +134,29 @@ inline void lazy_mutex_unsafe_unlock(lazy_mutex_handle &handle) {
+
+using recursive_mutex_handle = ::pthread_mutex_t;
+
+inline void recursive_mutex_init(mutex_handle &handle, bool checked = false) {
```
It so happens that `mutex_handle` and `recursive_mutex_handle` are the same here, but really these should take `recursive_mutex_handle`.

al45tair commented on include/swift/Threading/Impl/Pthreads.h:
```diff
@@ -130,6 +130,29 @@ inline void lazy_mutex_unsafe_unlock(lazy_mutex_handle &handle) {
+
+using recursive_mutex_handle = ::pthread_mutex_t;
+
+inline void recursive_mutex_init(mutex_handle &handle, bool checked = false) {
```
Similarly to the Linux implementation, these should probably take `recursive_mutex_handle` arguments.

mikeash committed:
Fix a bunch of RecursiveMutex code using plain mutex types.

al45tair committed:
[Threading] Add `_WIN64` conditional version of `ULONG_PTR`.

`ULONG_PTR` is defined slightly differently on `_WIN64` builds.

rdar://113898653

al45tair commented:
@swift-ci Please test Windows platform

mikeash commented:
@swift-ci please test

mikeash committed:
Fix concurrency-leaks.swift test to account for no more status lock record.

mikeash commented:
`embedded/concurrency-leaks.swift` fails because it checks for the lock record being allocated and destroyed. Removing those lines since that no longer happens.

mikeash commented:
Windows failed due to an unrelated issue on `main` which is now fixed, retrying.

mikeash commented:
@swift-ci please test windows platform

UncleMattHope commented on include/swift/ABI/Task.h:
```diff
@@ -298,7 +298,18 @@ class AsyncTask : public Job {
   /// Private storage for the use of the runtime.
   struct alignas(2 * alignof(void*)) OpaquePrivateStorage {
-    void *Storage[14];
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION && SWIFT_POINTER_IS_4_BYTES
```
Sorry for the late comment, I'm just slightly confused why there's a ``#if`` condition where both paths are identical in terms of text

mikeash replied:
Well spotted. The reason is, I messed it up. I have a fix going here: https://github.com/swiftlang/swift/pull/80447
